Add reorderable list component

alex-ju opened the pull request:
Add reorderable list component to be used for reordering attachments in Publications.

## Preview

### Desktop

<table>
<tr><th>JavaScript enabled</th><th>JavaScript disabled</th></tr>
<tr><td valign="top">

![reorder-js-enabled-desktop](https://user-images.githubusercontent.com/788096/75340136-cfe6e180-5889-11ea-94f0-4ec7302bb9eb.png)

</td><td valign="top">

![content-publisher dev gov uk_component-guide_reorderable_list_default_preview_](https://user-images.githubusercontent.com/788096/76445589-9059e280-63bd-11ea-9d5d-aa4cbf4ab19c.png)

</td></tr>
</table>

### Mobile
<table>
<tr><th>JavaScript enabled</th><th>JavaScript disabled</th></tr>
<tr><td valign="top">

![reorder-js-enabled-mobile](https://user-images.githubusercontent.com/788096/75340436-65827100-588a-11ea-8845-3f33ec4d517c.png)

</td><td valign="top">

![content-publisher dev gov uk_component-guide_reorderable_list_default_preview_(iPhone 6_7_8 Plus)](https://user-images.githubusercontent.com/788096/76445739-cbf4ac80-63bd-11ea-8c32-d74e534e0678.png)


</td></tr>
</table>

### Screen reader instructions
Examples using VoiceOver
<table>
<tr><th>JavaScript enabled</th><th>JavaScript disabled</th></tr>
<tr><td valign="top">

> Move Budget 2018 up, button, list 5 items.

</td><td valign="top">

> Order index for Budget 2019, edit text, list 5 items

</td></tr>
</table>

[Trello card](https://trello.com/c/H3T7JWtS)

benthorner commented on app/assets/stylesheets/components/_reorderable-list.scss:
```diff
@@ -0,0 +1,128 @@
+}
+
+.app-c-reorderable-list__description {
+  @include govuk-font(16, regular);
```
This confused me slightly, because we're making a 'global' font setting and then undoing it here. Would it be simpler to style the title and indices explicitly instead?

alex-ju replied:
The global font setting is at the list level and it's required to style the numbers – it also implicitly applies to list items. We're using this as an exception for the description indeed.

benthorner commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -8,10 +8,30 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
+  }
+
+  ReorderableList.prototype.setupResponsiveChecks = function () {
+    this.mql = window.matchMedia('(min-width: 40.0625em)')
```
Perhaps we could query the device instead? Especially as a narrow width window on my desktop is still usable for drag and drop.

alex-ju replied:
This is the value we use to detect the mobile breakpoint. We can store it into a const to give it name, but we're only using it once in the code.

kevindew replied:
Yeah I think something extra that tells us what this means would be useful. I was able to google a px converter and work out that it's probably 641px.

Maybe something like:
```
var tabletUp = (min-width: 40.0625em) // 641px ?
```

variable as Ben is suggesting is more for intent based documentation than re-use.

alex-ju replied:
Updated to store the value in a `tabletBreakpoint` variable, for intent-based documentation.

benthorner commented on app/assets/stylesheets/components/_reorderable-list.scss:
```diff
@@ -0,0 +1,128 @@
+
+.js-enabled {
+  .app-c-reorderable-list__actions {
+    .govuk-form-group {
```
This is a bit surprising. Could we refer refer to the group (the order index) by name somehow?

alex-ju replied:
It seems like I missed this comment. We can refer to the input by name, but not to the group (which is the outermost wrapper of the input component).

benthorner commented on app/assets/stylesheets/components/_reorderable-list.scss:
```diff
@@ -0,0 +1,128 @@
+    }
+  }
+
+  .app-c-reorderable-list__item:first-child .gem-c-button:first-of-type,
```
Same point here about having a way to refer to these buttons more explicitly.

alex-ju replied:
The advantage of using CSS pseudo-selectors in this way is that we don't have to add JavaScript to manipulate the DOM and update classes when reordering elements.

benthorner commented on app/assets/stylesheets/components/_reorderable-list.scss:
```diff
@@ -0,0 +1,128 @@
+    display: none;
+
+    @include govuk-media-query($from: desktop) {
+      display: inline-block;
```
Minor: do we need this if we're setting it for `.gem-c-button`?

alex-ju replied:
We need it to override the `display: none;` used until `$desktop` breakpoint, on L117.

benthorner commented on app/assets/stylesheets/components/_reorderable-list.scss:
```diff
@@ -0,0 +1,128 @@
+
+  .app-c-reorderable-list__item:first-child .gem-c-button:first-of-type,
+  .app-c-reorderable-list__item:last-child .gem-c-button:last-of-type {
+    display: none;
```
Could you help me understand what this is for?

alex-ju replied:
The up button for the first item and the down button for the last item are being hidden using different techniques on mobile compared to desktop. On mobile we use `display: none` as we don't want them to be shown, nor to occupy layout space (which helps us aligning them to the left). On desktop we use `visibility: hidden` which hides the button but takes up the same space as before which helps us keeping the up/down buttons aligned (more precisely the down button for the first item).

benthorner commented on app/assets/stylesheets/components/_reorderable-list.scss:
```diff
@@ -0,0 +1,128 @@
+    .gem-c-button:first-of-type {
+      margin-left: 0;
+
+      @include govuk-media-query($from: desktop) {
```
I'm struggling to interpret this sequence of styles. I think the purpose of this is to ensure we don't have a left margin when the buttons have wrapped onto their own line. Could we achieve the same thing more explicitly by only setting a `margin-left` on the `last-of-type` (and perhaps a `margin-right` on the content to the left)?

alex-ju replied:
Yes, it can be done that way too, not convinced it's more explicit thought since we'd only want the `margin-right` on the content if there are buttons on the right-hand side.

alex-ju commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -8,10 +8,30 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
+  }
+
+  ReorderableList.prototype.setupResponsiveChecks = function () {
+    this.mql = window.matchMedia('(min-width: 40.0625em)')
```
a `MediaQueryList` object. we need to store it so we can use it [in `checkMode` function, a few lines below](https://github.com/alphagov/content-publisher/pull/1819/files/67635252b5b937886efb2c3f6f2e04b0759142ed..6ede0ebd3576cad8ef30fdb56af579636ce9dd1e#diff-ddc0e2bfcf373620a17549ad2593fe83R30).

alex-ju replied:
Updated as suggested.

alex-ju commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -8,10 +8,30 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
       dragClass: "app-c-reorderable-list__item--drag"
     })
+
+    if (typeof window.matchMedia === 'function') {
```
This is our way of doing feature detection on media queries. If we want to be defensive, if not available we can disable the drag and drop indeed. Will update and record the decision.

alex-ju commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -7,17 +7,27 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
-    this.sortable = new window.Sortable.create(this.$module, {
-      chosenClass: "app-c-reorderable-list__item--chosen",
-      dragClass: "app-c-reorderable-list__item--drag"
+    this.sortable = new window.Sortable.create(this.$module, { // eslint-disable-line new-cap
```
Indeed, will rebase. Haven't got my linter set-up correctly.

kevindew replied:
Are you sure this should be a new call? looking at the docs the use this without new: https://github.com/SortableJS/Sortable#usage (It seems quite unusual to be calling new on what looks like a class method)

alex-ju replied:
If it's not a new call, multiple instances of this component on the same page will share the same instance of SortableJS – I assume we don't want this feature.

alex-ju replied:
As we checked together, the create function generates a new instance already, so I've updated our code to remove the `new` operator.

alex-ju commented on app/views/components/_reorderable_list.html.erb:
```diff
@@ -0,0 +1,38 @@
+        <%= tag.div class: "app-c-reorderable-list__actions" do %>
+          <%= render "govuk_publishing_components/components/input", {
+            label: {
+              text: sanitize("Order index<span class='govuk-visually-hidden'> for #{item[:title]}</span>")
```
Yes, `aria-label` is a good alternative – will have to update the component to allow it.

alex-ju commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -7,17 +7,27 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
 
   ReorderableList.prototype.start = function ($module) {
     this.$module = $module[0]
+    this.$upButtons = this.$module.querySelectorAll('.app-c-reorderable-list__actions button:first-of-type')
```
I would've normally used a `.js-reorderable-list-up-button` class for querying the elements via JavaScript (as we do in all components) but components from the library don't allow custom classes to be added.

alex-ju replied:
In this specific case, it won't work as the `:first-of-type` pseudo selector only works on different element types. What would be the advantage of using the `.gem-c-button` class instead?

Sure, if we're not in a rush with this will start again the conversation around allowing classes to be passed in for components.

alex-ju replied:
Opened https://github.com/alphagov/govuk_publishing_components/pull/1344 for this.

alex-ju commented:
> We need to think about the data this element will submit. Firstly, the `order_index` inputs will need to be kept in-sync when elements move on the page. Secondly, we need to know which index goes with each element. There are a couple of ways we could do this. I think the easiest option is for each item to have an `original_order_index[]`, so we can work out how the items have moved.

I was hoping that we can save users from some JavaScript being served and computations being done after each move by exposing a reversed attribution `input-name[new_index]=old_index.
I'll try and exemplify how I envisioned the order submission would work:

no-js: before reorder
```sh
reorderable-list[0]=1
reorderable-list[1]=2
reorderable-list[2]=3
```

no-js: after reorder (users manually replacing the 2 with 1 in the first input and vice-versa for the second)
```sh
reorderable-list[0]=2 #the second item will get index 0
reorderable-list[1]=1 #first item will get index 1
reorderable-list[2]=3 #third item remains the same
```

js enabled: before reorder (after script is being executed)
```
reorderable-list[]=1
reorderable-list[]=2
reorderable-list[]=3
```

js enabled: after reorder
```
reorderable-list[]=2
reorderable-list[]=1
reorderable-list[]=3
```

Now looking at it may work fine, but what I'm not capturing here is how do proceed when we have custom item IDs. I'm trying to figure out what a new set of inputs storing `original_order_index` would involve (both frontend and backend) – maybe we can catch-up on this.

> We seem to lose the focus states for the Up/Down buttons on click and on "Enter". I think there may be a general point there about indicating the element that was moved.

Great point. I've updated the code to preserve the focus.

> There's something weird going on with the drag/drop, which I can demo in person: sometimes the item 'jumps' downward and then returns to its current position.

Thanks for demo-ing this bug. I've updated the code to fix this.

> There's a slight inconsistency in the total space of the buttons for the first/last items. Probably not a big deal, but I thought I'd mention it, since we were concerned about it in the first commit.

Updated the code to fix these inconsistencies.

> My screen reader is seeing the first 'Down' button as a list of 5 buttons. I had a quick Google but can't immediately see the reason for this - I'm guessing it's reading down the 'column'.

It's a standard way for screen readers to let you know this is an element in a list of 5 items – valuable information when you're reordering.

> When using my screen reader I found the 'Move' help text confusing, since it sounds like it's part of the title. Can we make the label just `Move Up` and "`aria-describedby`" the title?

Replied in an inline comment below; I consider the `aria-label` suggestion as being more practical that `aria-describedby` – regardless it requires an update on the button component.

ChrisBAshton commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -34,5 +44,23 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
 
+  ReorderableList.prototype.onClickUpButton = function (e) {
+    var item = this.closest('.app-c-reorderable-list__item')
+    var previousItem = item.previousElementSibling
```
Nice - I hadn't come across these DOM properties before. 🎉

ChrisBAshton commented on app/views/components/_reorderable_list.html.erb:
```diff
@@ -0,0 +1,39 @@
+  items ||= []
+  data_attributes ||= {}
+  data_attributes[:module] = "reorderable-list"
+  input_name ||= "reorderable-list-#{SecureRandom.hex(4)}"
```
One other suggestion that has just occurred to me - should the component take an `error_message`/`error_items` property like [other components in govuk_publishing_components](https://github.com/alphagov/govuk_publishing_components/blob/master/app/views/govuk_publishing_components/components/_input.html.erb)?

I imagine if a user enters an unparseable input on one of the inputs, we should highlight the specific input in a red error state and explain what is wrong.

I'm a little hesitant to suggest this, as I know @benthorner was keen we don't bog ourselves down in validation. I suppose the alternative would be to just skip over any offending inputs and put them at the back of the queue, so to speak, though once we start adding that logic we're not far off the complexity we'd get from adding 'proper' validation.

alex-ju replied:
Good question about this. I didn't want to introduce it in the component before discussing how/if we're going to support this.

ChrisBAshton commented on app/views/components/_reorderable_list.html.erb:
```diff
@@ -0,0 +1,39 @@
+          <%= tag.p item[:description], class: "app-c-reorderable-list__description" %>
+        <% end %>
+        <%= tag.div class: "app-c-reorderable-list__actions" do %>
+          <%= render "govuk_publishing_components/components/input", {
```
Side note: is it worth using `<input type=”text” inputmode=”numeric” pattern="[0-9]*">` here as per https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/?

alex-ju replied:
Most probably very few/if nay people may benefit from this. Happy to add it if you think it's useful, but will have to update the input component to support these attributes.

alex-ju replied:
Opened https://github.com/alphagov/govuk_publishing_components/pull/1345 for this

alex-ju replied:
Opened https://github.com/alphagov/govuk_publishing_components/pull/1345 for this

ChrisBAshton commented:
One last comment from me!

Been thinking about how best to represent the ordering. I think the backend can be kept very simple (no need to pass a `original_order_index`, or indeed any index) and frontend needs to render only one input:

```erb
<% items = [{ id: 43, type: "FileAttachment"}, { id: 37, type: "FileAttachment" }] %>
<% # ^ example output from backend %>
<% items.each_with_index do |item, index| %>
  <input type="text" name="order_for_item_#{item.id}" value="#{index}" />
<% end %>
```

NoJS version: user can edit the value of the input, and submit the form. JS version: user can drag-and-drop (or use buttons) to move the element around; the JS should update the `value` of all inputs automatically. Then form submission would be something like:

```ruby
def reordered_items
  items.sort do |a, b|
    params["order_for_item_#{a.id}"].to_f <=> params["order_for_item_#{b.id}"].to_f
  end
end
```

Although it's nice submitting an input array (as that shows some relationship between the inputs), I think this approach is simpler as it means not having to maintain two separate arrays for original and new ordering.

benthorner commented:
@ChrisBAshton thanks for the suggestion. I can see a couple of problems we would encounter when we come to implementing it, which make me think we should still with the two-array approach:

   - Each item would be a different parameter. This means we would have to manually extract each one from the params by name, instead of a single array. It's also a bit of a smell that a group of related values aren't actually grouped, except by a common name.
   - The ID of featured attachments is not unique, since it depends on a combination of the database ID and the type of attachment model we're referring to, as mentioned [on the card about reordering](https://trello.com/c/7NXYdxZV/1566-user-can-reorder-featured-attachments). We did explore this approach, but believe it's simpler to focus the data around the order, not IDs.

ChrisBAshton commented:
We wouldn't necessarily have to use `.id` - we could pass some combination of id and attachment type to the frontend. So long as it's globally unique - and reversible - the frontend doesn't need to know how we've arrived at the ID, we're only using it ephemerally and can change our implementation of the ID over time if needed.

I do agree about the code smell 👍 I've thought more about your proposed implementation - I think what you're proposing is this:

```erb
<%
  items = [
    { id: 43, type: "FileAttachment", order_index: 1, original_order_index: 1 },
    { id: 37, type: "FileAttachment", order_index: 2, original_order_index: 2 }
  ] %>
<% # ^ example output from backend %>

<% items.each do |item| %>
  <input type="hidden" name="original_order[]" value="#{item.original_order_index}" />
  <input type="text" name="new_order[]" value="#{item.order_index}" />
<% end %>
```

...on form submission, we'd have something like:

```ruby
def reordered_items
  # params["original_order"] # ["1", "2", "3"] # I don't think we'll need to use this!
  # params["new_order"] # ["2", "1", "3"]

  params["new_order"].map do |old_index|
    items[old_index.to_i]
  end
end
```

This, too, can be simplified I think 🤔 

```erb
<% items = [{ id: 43, type: "FileAttachment"}, { id: 37, type: "FileAttachment" }] %>
<% # ^ example output from backend %>
<% items.each_with_index do |item, index| %>
  <input type="text" name="new_order[]" value="#{index}" />
<% end %>
```

Form submission:

```ruby
def reordered_items
  # params["new_order"] # ["2", "1", "3"]

  params["new_order"].map do |old_index|
    items[old_index.to_i]
  end
end
```

Either way, I don't think the backend needs to pass additional array(s) of orderings. They can be inferred by the order of the `items` array and some clever manipulation of the form DOM.

ChrisBAshton commented:
Chatted IRL with Ben, I now see the issue 😅 

```html
* ('foo' param added to disambiguate each input, since their values can change)

<input foo="bar" val="1">
<input foo="baz" val="2">
<input foo="tom" val="3">

<!-- ...reordered with JavaScript -->
<input foo="tom" val="3">
<input foo="bar" val="1">
<input foo="baz" val="2">

<!-- Submit -->
...would give form params of: [3, 1, 2]
...which could be mapped to: [tom, bar, baz]
...great!

But the noJS version wouldn't use ordering, it would involve changing the values:

<input foo="bar" val="2">
<input foo="baz" val="3">
<input foo="tom" val="1">

<!-- Submit -->
...would give form params of: [2, 3, 1]
...which would be mapped to: [baz, tom, bar]
...and NOT [tom, bar, baz] as we wanted!
```

Not sure why that took so long to click!

And to clarify, the suggested implementation is:

```erb
<% items = [{ id: 43, type: "FileAttachment"}, { id: 37, type: "FileAttachment" }] %>
<% # ^ example output from backend %>

<% items.each_with_index do |item, index| %>
  <input type="hidden" name="original_order[]" value="#{index}" />
  <input type="text" name="new_order[]" value="#{index}" />
<% end %>
```

(i.e. we will not pass any extra params from the backend. Rather, it is the frontend that will pass the ordering params to the backend).

alex-ju commented on app/views/components/_reorderable_list.html.erb:
```diff
@@ -0,0 +1,38 @@
+      <%= tag.div class: "app-c-reorderable-list__wrapper" do %>
+        <%= tag.div class: "app-c-reorderable-list__content" do %>
+          <%= tag.p item[:title], class: "app-c-reorderable-list__title" %>
+          <%= tag.p item[:description], class: "app-c-reorderable-list__description" %>
```
Most of the app-level components in Content Publisher use helper functions to generate markup. Because we don't have a strict set of rules around how markup should be generated components in govuk_publishing_components are written in a mix of ways. If you pick [the accordion component](https://github.com/alphagov/govuk_publishing_components/blob/master/app/views/govuk_publishing_components/components/_accordion.html.erb) you'll find the same helpers (aside from one tag, to be fair). I have a note to start discussions around this within the community but it got deprioritised a few times in the past year.

I personally prefer using helpers as they tend to be easier to read and more forgiving when not providing options (by not generating empty attributes, which otherwise would require lots of conditional statements).

alex-ju commented on app/views/components/docs/reorderable_list.yml:
```diff
@@ -0,0 +1,30 @@
+
+  * be keyboard focusable
+  * inform the user about which item they operate on
+  * preserve focus after interacting with them
```
All of our components are meant to be build using progressive enhancement.

alex-ju commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -12,6 +12,26 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
     })
+
+    if (typeof window.matchMedia === 'function') {
+      this.setupResponsiveChecks()
```
Hmm.. this doesn't apply something really, it just adds a listener to check for a breakpoint. Probably `checkMode` could be renamed as it doesn't only check the media query, but also disable the feature.

alex-ju commented:
@ChrisBAshton Thank you for the review and for trying to find a way to only use one input. I've been 'seduced' by the same drive to reduce the number of inputs, but it seems like it just doesn't fit with the requirements.

benthorner commented:
@alex-ju I think @kevindew is planning to comment, so this is just a pre-emptive warning. After talking with @kevindew, we think there's a better approach we can take to solving the ordering approach in the backend - sorry for so many iterations! The quick summary is:

   - introduce an extra UUID into the ordering data we store (and rename `id`/`type` to `attachment_X`)
   - have a single params *hash* that's set like `attachments[uuid]=<order_index>` to represent the reordering

So each item/attachment still needs to arguments set, but we're proposing to use them differently.

kevindew commented:
Ah yes it was taking me a while to get the notes down. I've just had a chat with Ben about the parameters here and I'm summarising it here. 

## Two parameter fields

We talked about the need for two parameter fields, the working understanding was that we would need two input fields for example:

```
<div class="reorderable-attachment">
  <input type="hidden" name="previous_ordering[]" value="0" />
  <input type="text" name="new_ordering[]" value="2" />
</div>
```

Where this would look up the item in the file attachment ordering at item 0 and then move that item to be at position 2.

We've identified that this could be resolved with passing the id in as part of the params. For example:

```
<div class="reorderable-attachment">
  <input type="text" name="ordering[0]" value="2" />
</div>
```

Where the ordering hash could be looped through and the item at 0 could then be moved to position 2.

## Using the ordering as an id

We then talked about the risks in using the ordering as index to look up which items should have its position moved. We talked in a particular about a concurrent editing situation where for example: An editor asks a slack room for someone to move a CSV to the top of the list of featured attachments. If both Kevin & Ben load the page before the other has submitted their edits the first person (Ben) to submit the form will achieve the ordering as expected, whereas when Kevin submits it the data submitted on current ordering will be out of date and will undo the changes Ben made.

To resolve this we talked about what could be used as an id and came up with perhaps using a UUID. The effect of this could be changing the featured attachment array to something like:

```
[
  { id: "696d2cfb-b97a-479f-952a-abcef0c1c88e",
    attachment_type: "FileAttachment",
    attachment_id: 123 },
  { id: "217ab88f-74a9-41eb-ab47-428f54461c02",
    attachment_type: "HtmlAttachment",
    attachment_id: 456 }
]
```

and then changing the input approach to something like:

```
<div class="reorderable-attachment">
  <input type="text" name="ordering[696d2cfb-b97a-479f-952a-abcef0c1c88e]" value="1" />
</div>
```

So I'm guessing we'd end up with something along these lines to do the ordering (where the infinity is used to push any items missed from submission to the end of the list and doing nothing can have them at the start): 

```
id_1, id_2, id_3 = 3.times.map { SecureRandom.uuid }
file_attachments = [{ id: id_1 }, { id: id_2 }, { id: id_3 }]
submission = { id_1 => "1", id_2 => "0" }
ordering = submission.sort { |a, b| a.last.to_i <=> b.last.to_i }.map(&:first)
sorted = file_attachments.sort_by { |a| ordering.index(a[:id]) || Float::INFINITY }
```

alex-ju commented:
> To resolve this we talked about what could be used as an id and came up with perhaps using a UUID.

Thank you for the comment @kevindew. Coping with concurrency was what I was thinking when I referred to 'custom item IDs' above, and in previous discussions. What I wasn't sure about was how permissive the index is in terms of character types (e.g `inputname[696d2cfb-b97a-479f-952a-abcef0c1c88e]` which contains `-`) but I imagine it's flexible enough. I updated the code to support custom indexes again, only that now both inputs will share the same index.

kevindew commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -0,0 +1,84 @@
+    if (this.mql.matches) {
+      this.sortable.option('disabled', false)
+    } else {
+      this.sortable.option('disabled', true)
```
You can probably shorten this to `this.sortable.option('disabled', this.mql.matches)`

alex-ju replied:
Indeed, only the other way around (`!this.mediaQueryList.matches`) – I wasn't sure if the intent will remain clear. Let me know if you want me to shorten it like this.

alex-ju replied:
Replaced this logic with a one-liner

kevindew commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -0,0 +1,84 @@
+    }
+    // if triggered by keyboard preserve focus on button
+    if (e.detail === 0) {
+      e.target.focus()
```
There seems to be a case here where if you move an item up to the top with keyboard this will apply a focus to a hidden element - and after that the tab index seems a little strange. Same situation happens on down too.

alex-ju replied:
Yes, it's an edge case that I haven't captured in the code. We will need to add more logic to cope with these two cases (first visible up button and last visible down button).

kevindew commented on app/assets/javascripts/components/reorderable-list.js:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  ReorderableList.prototype.updateOrderIndexes = function () {
+    var $orderInputs = this.$module.querySelectorAll('.app-c-reorderable-list__actions input')
```
Is there a more precise selector we can use on these?

alex-ju replied:
We can, but given we control the markup within the component I wouldn't add complexity to this query. What's the concern around it?

kevindew commented on app/views/components/_reorderable_list.html.erb:
```diff
@@ -0,0 +1,39 @@
+  items ||= []
+  data_attributes ||= {}
+  data_attributes[:module] = "reorderable-list"
+  order_input_name ||= "reorderable-list-order-#{SecureRandom.hex(4)}"
```
Since names don't have to be unique I don't think we really benefit from the random hex here

alex-ju replied:
Yep, a leftover. More than that, I think this attribute should be required (no default value provided) so I removed this line.

kevindew commented on app/views/components/docs/reorderable_list.yml:
```diff
@@ -0,0 +1,54 @@
+      order_input_name: new_order
+      items:
+        - title: "Budget 2018"
+          description: "PDF, 2.56MB, 48 pages"
```
I've talked to Simon that doing these trio of attributes is not going to be very easy as this is data that is wrapped up in the attachment component. So we talked about simplifying them. It shouldn't matter to you much since we're just passing a string in here though.

alex-ju commented on app/assets/stylesheets/components/_reorderable-list.scss:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  @include govuk-media-query($from: desktop) {
+    cursor: move;
```
Good spot. Moved it in the `js-enabled` block.

alex-ju commented:
> I've added a bunch of comments. I also was a bit confused about whether some of the existing ones were pending response/resolution?

Thank you, @kevindew. I tried to reply to them all.

alex-ju commented:
@ChrisBAshton, sure I will. I don't expect to get this merge anytime sooner, as it needs refactoring after the requested changes in govuk_publishing_components.

alex-ju commented:
Updated screenshots, updated the code to profit from the changes introduced in `govuk_publishing_components` for [buttons](https://github.com/alphagov/govuk_publishing_components/pull/1344) and [inputs](https://github.com/alphagov/govuk_publishing_components/pull/1345). Unfortunately, I've only got support in using custom classes for `js-` selectors (which cleans the JavaScript code a bit), but not for styling (which leaves us with things like `gem-c-button:first-of-type` in our code). I'll try and raise the custom classes topic again in community discussions. I hope I addressed and/or reply to all queries and suggestions.
